### DISCUSSION TITLE
Dotted Border around Buttons

### DISCUSSION THREAD

**Discussion by Redflashx12** [fltk-rs/fltk-rs#1460] [Q&A] [answered]
Hey, 
I'm new to FLTK and I have a small issue which I am having trouble with debugging. In our project, we have a multi-threaded setup with the GUI having its own thread. Now I have three buttons here and currently, they have this dotted border around them:

![image](https://github.com/fltk-rs/fltk-rs/assets/43482469/0b0f70fc-305d-4cdd-bee1-1f5233dee335)

First I tried to look it up and found [this thread](https://fltk.easysw.narkive.com/wmTOMH48/dotted-line-around-last-widget) mentioning the visible_focus and I tried disabling it globally and/or for each button, but neither worked.

Then I tried to remove the functionality from the buttons and test the layout independently to see if it appeared, but it didn't. 
What I did manage to debug was that if the window is not redrawn in the run method, at the beginning at least they are drawn as intended:
![image](https://github.com/fltk-rs/fltk-rs/assets/43482469/20d85295-2851-485b-b0ae-cd1472761b70)

However, if I click on any button, it gets back to the state with the dotted border, e.g. here:
![image](https://github.com/fltk-rs/fltk-rs/assets/43482469/8823269b-7c86-4e92-bad6-d8fe9ca1976d)

The 'run' method calls the 'layout' function and gets the frame, app, etc. and draws project-related images and bounding boxes and redraws the frames and window at the end (everything in a while app.wait() loop), but I also tried to reproduce that in a separate project with a more simple setup but I also didn't manage to reproduce it.

Would appreciate any help or tips. I'm a bit lost. If you need any other information, let me know.

The code for the layout (with just the start/wait + stop button) is this:

```
use std::thread;
use std::time::Duration;
use fltk::{app, button::Button, draw, frame, frame::Frame, group, group::Flex, prelude::*, text, window, window::Window};
use fltk::app::App;
use fltk::enums::{Align, Color, FrameType};
use fltk::text::TextDisplay;

pub fn layout(
)-> (Frame, Window, App, TextDisplay, Button) {
    let app = app::App::default().with_scheme(app::Scheme::Base);

    let wind_width = 1920;
    let wind_height = 1080;
    let mut wind =
        window::Window::new(0, 0, wind_width, wind_height, "WeSort.AI Core").center_screen();
    wind.set_color(Color::White);

    let mut pack = group::Pack::new(0, 0, wind_width, wind_height, "");
    pack.set_spacing(0);
    pack.set_type(group::PackType::Vertical);

    let mut top_bar = group::Pack::new(0, 0, wind_width, 32, "");
    top_bar.set_spacing(0);
    top_bar.set_type(group::PackType::Horizontal);
    top_bar.end();

    // Add a start/stop button at the top left
    let wait__label = "@+6||";
    let start_label = "@+6>";
    let stop_label = "@+3square";
    let problem_indicator_label = "@-1circle";

    let mut start_wait_button = Button::new(0, 0, 32, 32, "");
    start_wait_button.set_color(Color::White);
    start_wait_button.set_frame(FrameType::RFlatBox);
    start_wait_button.set_label(wait__label);
    start_wait_button.set_label_size(16);
    start_wait_button.set_label_color(Color::Green);
    start_wait_button.set_down_frame(FrameType::RFlatBox);

    let mut has_start_button = false;

    start_wait_button.set_callback(move |start_wait_button| {
        if has_start_button == false {
            has_start_button = true;

            start_wait_button.set_label(start_label);
            start_wait_button.set_frame(FrameType::RFlatBox);
            start_wait_button.set_down_frame(FrameType::RFlatBox);
        } else {
            has_start_button = false;

            start_wait_button.set_label(wait__label);
            start_wait_button.set_frame(FrameType::RFlatBox);
            start_wait_button.set_down_frame(FrameType::RFlatBox);
        }
    });

    top_bar.add(&start_wait_button);

    let mut stop_button = Button::new(40, 0, 32, 32, "");
    stop_button.set_color(Color::White);
    stop_button.set_frame(FrameType::RFlatBox);
    stop_button.set_label(stop_label);
    stop_button.set_label_size(16);
    stop_button.set_label_color(Color::Red);
    stop_button.set_down_frame(FrameType::RFlatBox);

    top_bar.add(&stop_button);

    let mut column = group::Flex::default_fill().column();

    let frame = {
        let mut col2 = group::Flex::default().with_label("Live View   ").column();
        let mut frame = frame::Frame::new(0, 300, wind_width, wind_height, "");
        col2.add(&mut frame);
        col2.end();
        frame
    };

    let frame_pixel_height = 1000;

    let text = {
        let col1 = group::Flex::default().with_label("Dashboard   ").column();
        let text = text::TextDisplay::new(
            0,
            (frame_pixel_height / 2) as i32,
            wind_width,
            wind_height - frame_pixel_height as i32 / 2,
            "",
        );
        col1.end();
        text
    };

    top_bar.end();
    column.end();
    pack.add(&top_bar);
    pack.add(&column);
    pack.end();
    pack.set_size(wind_width, wind_height);

    wind.make_resizable(true);
    wind.end();
    wind.show();

    (frame, wind, app, text, start_wait_button)
}
```

The run method is kind of set up like this, but drawing something else, but when I'm using this simplified setup everything works as intended:

```
fn main() {
    let (mut frame, mut wind, app, mut text, mut button) = layout();

    while app.wait() {

        frame.draw(move |f| {
            draw::draw_rect_fill(0, 50, f.width(), f.height()-50, Color::Black);
        });
        frame.redraw();

        wind.redraw();
        thread::sleep(Duration::from_millis(10));
    }
}
```

**Comment by MoAlyousef** [comment]
Hi

This looks more like an artifact in label drawing. The focus rect that you get when you have visible focus would outline the button’s frame, not the button’s label. 

So I’m not sure what might cause this. 

What OS are you experiencing this on?

If linux, X11 or wayland? Do you use any specific features when building fltk-rs? What version of fltk-rs?

If you’re able to create a minimal repro which produces the artifact, it would simplify debugging things. 

Unrelated, but you shouldn’t be setting the draw method in your app.wait() loop. It unnecessarily allocates a new draw callback with each iteration.

**Reply by Redflashx12** [reply]
Since our target machine is supposed to run on Windows, I'm developing it on Windows 11.

The fltk version is 1.4.12, features are only ["fltk-bundled"]

My colleague who has NixOS with Wayland doesn't have this issue.

And thanks for the unrelated note. 👍 

Will try to reproduce this with a minimal repro.

**Reply by Redflashx12** [reply, marked as answer]
Okay, just found the issue. My colleague implemented the other part and he was changing the draw::set_line_style to Dash for the bounding box and dotted for some other visualization. Now, he didn't reset it back to solid afterwards which is mentioned as a warning in the documentation of the method.

The weird part is however that he did not have any issue with this, but on Windows, I did have this problem. Probably some implementation details. ❓

Thanks for the help :) Appreciate it.
